File: Java/MagicSquare.java
import java.util.*;
import java.util.regex.*;
import java.text.*;
import java.math.*;
import java.awt.geom.*;

public class MagicSquare
{
	private int ind(int[] square, int i, int j)
	{
		return square[i*3+j];
	}
	
	public int missing(int[] square)
	{
	for (int i = 0; i < 9; i++) if (square[i] == -1) square[i]= 0;
		int sum1 = ind(square, 0, 0)+ind(square,0,1)+ind(square,0,2);
		int sum2 = ind(square, 1, 0)+ind(square, 1, 1)+ind(square, 1, 2);
		int sum3 = ind(square, 2, 0)+ind(square, 2, 1)+ind(square, 2, 2);
		
		int sum = 0;
		if (sum1 == sum2) 
			sum = sum1;
		else if (sum1 == sum3)
			sum = sum1;
		else 
			sum = sum2;
			
		for (int r = 0; r < 3; r++)
		{
			int s = ind(square, r, 0)+ind(square, r, 1) + ind(square, r, 2);			
			if (s != sum) return sum-s;
			s = ind(square, 0, r)+ind(square, 1, r) + ind(square, 2, r);			
			if (s != sum) return (sum-s);		
		}
		int s = ind(square, 0, 0)+ind(square, 1, 1) + ind(square, 2, 2);			
		if (s != sum) return (sum-s);			
		s = ind(square, 0, 2)+ind(square, 1, 1) + ind(square, 2, 0);
		if (s != sum) return (sum-s);	
		
		return 0;		
	}
	
	public static void main(String[] args)
	{
		long time;
		int answer;
		boolean errors = false;
		int desiredAnswer;
		
		time = System.currentTimeMillis();
		answer = new MagicSquare().missing(new int[]{ 8, 1, 6, 3, 5, -1, 4, 9, 2 });
		System.out.println("Time: " + (System.currentTimeMillis()-time)/1000.0 + " seconds");
		desiredAnswer = 7;
		System.out.println("Your answer:");
		System.out.println("\t" + answer);
		System.out.println("Desired answer:");
		System.out.println("\t" + desiredAnswer);
		if (answer != desiredAnswer)
		{
			errors = true;
			System.out.println("DOESN'T MATCH!!!!");
		}
		else
			System.out.println("Match :-)");
		System.out.println();
		time = System.currentTimeMillis();
		answer = new MagicSquare().missing(new int[]{ -1, 1, 6, 3, 5, 7, 4, 9, 2 });
		System.out.println("Time: " + (System.currentTimeMillis()-time)/1000.0 + " seconds");
		desiredAnswer = 8;
		System.out.println("Your answer:");
		System.out.println("\t" + answer);
		System.out.println("Desired answer:");
		System.out.println("\t" + desiredAnswer);
		if (answer != desiredAnswer)
		{
			errors = true;
			System.out.println("DOESN'T MATCH!!!!");
		}
		else
			System.out.println("Match :-)");
		System.out.println();
		time = System.currentTimeMillis();
		answer = new MagicSquare().missing(new int[]{ 5, 15, 13, 19, 11, 3, 9, 7, -1 });
		System.out.println("Time: " + (System.currentTimeMillis()-time)/1000.0 + " seconds");
		desiredAnswer = 17;
		System.out.println("Your answer:");
		System.out.println("\t" + answer);
		System.out.println("Desired answer:");
		System.out.println("\t" + desiredAnswer);
		if (answer != desiredAnswer)
		{
			errors = true;
			System.out.println("DOESN'T MATCH!!!!");
		}
		else
			System.out.println("Match :-)");
		System.out.println();
		
		if (errors)
			System.out.println("Some of the test cases had errors :-(");
		else
			System.out.println("You're a stud (at least on the test data)! :-D ");
	}
}

//Powered by [KawigiEdit]

File: Misc/MagicSquare.cpp
#include <vector>
#include <list>
#include <map>
#include <set>
#include <deque>
#include <stack>
#include <bitset>
#include <algorithm>
#include <functional>
#include <numeric>
#include <utility>
#include <sstream>
#include <iostream>
#include <iomanip>
#include <cstdio>
#include <cmath>
#include <cstdlib>
#include <ctime>

using namespace std;

class MagicSquare {
public:
	int missing(vector <int> s) 
	{
		int a[3][3];
		int mr, mc, i = 0;
		for (int r = 0; r < 3; r++)
		{
			for (int c = 0; c < 3; c++)
			{
				a[r][c] = s[i];
				if (s[i] == -1)
				{
					mr = r;
					mc = c;
				}
				i++;
			}
		}
		int x = (mr+1)%3;
		int sum = a[x][0]+a[x][1]+a[x][2];
		int sum2 = a[mr][0]+a[mr][1]+a[mr][2]+1;
		return sum-sum2;
	}
};


//Powered by [KawigiEdit]

double test0() {
	int t0[] = { 8, 1, 6, 3, 5, -1, 4, 9, 2 };
	vector <int> p0(t0, t0+sizeof(t0)/sizeof(int));
	MagicSquare * obj = new MagicSquare();
	clock_t start = clock();
	int my_answer = obj->missing(p0);
	clock_t end = clock();
	delete obj;
	cout <<"Time: " <<(double)(end-start)/CLOCKS_PER_SEC <<" seconds" <<endl;
	int p1 = 7;
	cout <<"Desired answer: " <<endl;
	cout <<"\t" << p1 <<endl;
	cout <<"Your answer: " <<endl;
	cout <<"\t" << my_answer <<endl;
	if (p1 != my_answer) {
		cout <<"DOESN'T MATCH!!!!" <<endl <<endl;
		return -1;
	}
	else {
		cout <<"Match :-)" <<endl <<endl;
		return (double)(end-start)/CLOCKS_PER_SEC;
	}
}

double test1() {
	int t0[] = { -1, 1, 6, 3, 5, 7, 4, 9, 2 };
	vector <int> p0(t0, t0+sizeof(t0)/sizeof(int));
	MagicSquare * obj = new MagicSquare();
	clock_t start = clock();
	int my_answer = obj->missing(p0);
	clock_t end = clock();
	delete obj;
	cout <<"Time: " <<(double)(end-start)/CLOCKS_PER_SEC <<" seconds" <<endl;
	int p1 = 8;
	cout <<"Desired answer: " <<endl;
	cout <<"\t" << p1 <<endl;
	cout <<"Your answer: " <<endl;
	cout <<"\t" << my_answer <<endl;
	if (p1 != my_answer) {
		cout <<"DOESN'T MATCH!!!!" <<endl <<endl;
		return -1;
	}
	else {
		cout <<"Match :-)" <<endl <<endl;
		return (double)(end-start)/CLOCKS_PER_SEC;
	}
}

double test2() {
	int t0[] = { 5, 15, 13, 19, 11, 3, 9, 7, -1 };
	vector <int> p0(t0, t0+sizeof(t0)/sizeof(int));
	MagicSquare * obj = new MagicSquare();
	clock_t start = clock();
	int my_answer = obj->missing(p0);
	clock_t end = clock();
	delete obj;
	cout <<"Time: " <<(double)(end-start)/CLOCKS_PER_SEC <<" seconds" <<endl;
	int p1 = 17;
	cout <<"Desired answer: " <<endl;
	cout <<"\t" << p1 <<endl;
	cout <<"Your answer: " <<endl;
	cout <<"\t" << my_answer <<endl;
	if (p1 != my_answer) {
		cout <<"DOESN'T MATCH!!!!" <<endl <<endl;
		return -1;
	}
	else {
		cout <<"Match :-)" <<endl <<endl;
		return (double)(end-start)/CLOCKS_PER_SEC;
	}
}

int main(int argc, char **argv) {
	double time;
	bool errors = false;
	
	if (argc < 2 || atoi(argv[1]) == 0) {
		time = test0();
		if (time < 0.0)
			errors = true;
	}
	
	if (argc < 2 || atoi(argv[1]) == 1) {
		time = test1();
		if (time < 0.0)
			errors = true;
	}
	
	if (argc < 2 || atoi(argv[1]) == 2) {
		time = test2();
		if (time < 0.0)
			errors = true;
	}
	
	if (!errors)
		cout <<"You're a stud (at least on the example cases)!" <<endl;
	else
		cout <<"Some of the test cases had errors." <<endl;
}

